### ISSUE TITLE
Custom keywords support to be able to use Roslyn for "C#-like" DSLs

### ISSUE THREAD

**Issue by rickardp** [dotnet/roslyn#27841]
**Background**
I've successfully been using Roslyn to create DSLs of varying complexity. Usually, I want to keep the expression and statement syntax essentially identical to C# and work with the Roslyn syntax tree to generate the code I want (for example, I have created a C#-like shader language based on Roslyn that transpiles to HLSL/GLSL).

Another example of how I have used this is to create a serialization format where I allow C# expressions in some contexts, and where C# types are first class citizens. Using Roslyn for this turned out to be very easy and equally powerful.

While lower levels of the C# syntax tree is usually good as is, I tend to want to change the upper levels, such as whether to allow usings, namespace declarations, and so on (this is easy to do currently), but I would also like to change for example the class declaration syntax (in the above example, I would have liked to replace "class" with "shader").

**TL;DR**
Turns out, this is actually very easy to implement in Roslyn with only a small change to the lexer (https://github.com/rickardp/roslyn/commit/681f20a28415974d57cc98e9e34cb3d5a57e0a59). I don't know if this has been considered and rejected before, but I could not find anything relevant. What I have added is the possibility to add synonyms for built-in keywords. It is then up to me as a user of Roslyn to understand and do something intelligent with the result, but this way at least I get correct syntax trees.

This way the changes stay minimal, as it can only use already existing syntactic concepts, not invent new ones. I think this is reasonable as it would be hard to keep the current level of performance otherwise (compare with generated parsers for example). I also believe a lot of very powerful concepts can piggyback on existing syntax.

A more elegant solution would be to introduce new token types for places where it is suitable to create new keywords (e.g. a “CustomTypeKeyword” token instead of a ClassKeyword token), but it would not provide anything more in terms of functionality.

I realize this might be slightly beyond the scope of Roslyn, but I also believe this would add a powerful feature for developers of C# DSLs / syntactic sugar.

**Comment by gafter** [comment]
We do not intend to add any features to make Roslyn a toolbox for building compilers.

**Comment by rickardp** [comment]
I see, and I understand this point. Is there any hope for the lexer API ever opening up so that the application can control the lexing part and tree parsing part independently? Or is that also beyond the scope for Roslyn?

Apart from my use case above, this could be useful if you wanted to add syntactic sugar, for example in an analyzer. This way it would also be trivial for language servers based on Roslyn to do the right thing w.r.t. syntax highlighting and such, by just pulling the right NuGet package in the project. (Assuming it would be possible, API-wise).

Or is extending the C# syntax in an application something you would prefer not to encourage? (i.e. by making it harder to do)

**Comment by CyrusNajmabadi** [comment]
>  Is there any hope for the lexer API ever opening up so that the application can control the lexing part and tree parsing part independently? Or is that also beyond the scope for Roslyn?

You can already do the former.  It's possible to just do the lexing by calling SyntaxFactory.ParseTokens.  However, there's no way to do the latter.  And, generally, this is out of scope for Roslyn.  Roslyn isn't a toolkit for building compilers.  Roslyn is the C# and VB compilers and language services.

> Or is extending the C# syntax in an application something you would prefer not to encourage? 

There is no desire by the team to make this possible.

> (i.e. by making it harder to do)

There's been no explicit effort (that i'm aware of) to go out of our way to make this harder.  It's just not something that is considered something that roslyn wants to do.  So no effort has been made to make this possible.  

--

Now on a personal level, i do not think it's a good idea.  This would just fragment the language into dialects that would then not interoperate.
